server/testdriver/core/src/main/java/org/infinispan/server/test/core/rollingupgrade/RollingUpgradeHandler.java

```diff
@@ -0,0 +1,229 @@
+               handler.logConsumer.accept("Shutting down 1 node from version: " + versionFrom);
+               int nodeId = nodeCount - i - 1;
+               String volumeId = handler.fromDriver.volumeId(nodeId);
+               handler.fromDriver.stop(nodeId);
+
+               handler.currentState = STATE.REMOVED_OLD;
+
+               if (!ensureServersWorking(handler, nodeCount - 1)) {
+                  handler.logConsumer.accept("Servers are: " + Arrays.toString(manager.getServers()));
+                  throw new IllegalStateException("Servers did not shut down properly within 30 seconds, assuming error");
+               }
+
+               handler.logConsumer.accept("Starting 1 node to version " + versionTo);
+               if (handler.toDriver == null) {
+                  handler.toDriver = handler.startNode(true, 1, nodeCount, site1Name,
+                        configuration.jgroupsProtocol(), volumeId);
+               } else {
+                  handler.toDriver.startAdditionalServer(nodeCount, volumeId);
+               }
+
+               handler.currentState = STATE.ADDED_NEW;
+
+               if (!ensureServersWorking(handler, nodeCount)) {
+                  handler.logConsumer.accept("Servers are only: " + Arrays.toString(manager.getServers()));
+                  throw new IllegalStateException("Servers did not cluster within 30 seconds, assuming error");
+               }
+            }
+
+            handler.currentState = STATE.NEW_RUNNING;
+         }
+      } catch (Throwable t) {
+         handler.currentState = STATE.ERROR;
+         configuration.exceptionHandler().accept(t, handler);
+         throw t;
+      } finally {
+         handler.cleanup();
+      }
+   }
+
+   private static boolean ensureServersWorking(RollingUpgradeHandler handler, int expectedCount) throws InterruptedException {
+      long begin = System.nanoTime();
+      while (TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - begin) < handler.configuration.serverCheckTimeSecs()) {
+         handler.logConsumer.accept("Checking to ensure cluster formed properly, expecting " + expectedCount + " servers");
+         if (handler.configuration.isValidServerState().test(handler)) {
+            return true;
+         }
+         Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+      }
+      handler.logConsumer.accept("Cluster state check timed out after " + handler.configuration.serverCheckTimeSecs() +
+            " secs, check for possible issues");
+      return false;
+   }
+
+   private void cleanup() {
+      if (fromDriver != null) {
+         fromDriver.stop(configuration.fromVersion());
+      }
+      if (toDriver != null) {
+         toDriver.stop(configuration.toVersion());
+      }
+   }
+
+   public RemoteCacheManager createRemoteCacheManager() {
+      TestUser user = TestUser.ADMIN;
+
+      String hotrodURI = "hotrod://" + user.getUser() + ":" + user.getPassword() + "@" + fromDriver.getServerAddress(0).getHostAddress() + ":11222";
```

Comment:
Should we use the configuration builder here? Should we extract the port from the InetAddress and fallback to 11222?

Reply:
Currently we don't override the port, so it will always be 11222. Honestly, this is just here to get something working which is very simple and replace it with something nicer as we need to.